Futuristic eqp

lawben opened the pull request:
Let's check what still needs to be done

lanice committed:
Projection::create_column only bypass column if all Expressions are of type ExpressionType::Column (#629)

* Bypass create_column in Projection only if all ColumnExpressions are columns.

* Format.

* Remove comment.

* Rename only_column_expression to reuse_column_from_input.

lawben committed:
Fix console error message for single statement queries (#632)

Commit:
Playground using namespace opossum (#635)

* Playground using namespace opossum

* retrigger jenkins

Commit:
Printing for PQPs (#634)

* Current state

* Current State

* current state

* current state

* format and a happy happy lint

* all build

* Update abstract_operator.cpp

Bensk1 committed:
Add missing include (#636)

Commit:
TPCH Benchmark Scheduler and MVCC  (#638)

* Initial commit

* Spelling

* Spelling

Commit:
Remove Chunk::use_mvcc_columns_from() (#639)

Commit:
RuleBatches (#637)

lawben committed:
Change defualt TpchDbGenerator chunk size (#642)

mrks committed:
Hash-Join: Make histogram creation faster (#647)

Bensk1 committed:
Fix tests relying on filenames (#641) (#643)

* Fix tests relying on filenames

* Move to test directory

* Create random uuids

* Handle all this with std::filesystem, add cmake magic

* Formatting

* Add nolint

* Adjust Jenkinsfile

* Clang linux

* Add comment

* Try c++17 over 1z

* Try linker filesystem clang fix

* Fu

* Revert "Fu"

This reverts commit 78fb084bb6077c2c5ebb07d02892de227252f94f.

* Fix global static

* Fufu

* Try

* sfpidjsdpfh

* Add library to systemTest

* Fix optional

* Minor

* Format

* FU

* Adjust directory names

* Decapitalize

* Move includes

* Remove testing utils

* K

* Cleanup creation

* Jenkinsfile

* coverage script

* Check for badge==true

* Clarify error message

* Clarify error message

* Remove test data initially

lawben committed:
Fix column order in multi-table FROM (#640)

* Fix column order in multi tabel FROM

* Upps. Shame on you guys for not noticing in the PR

* Change table order to bypass broken join ordering

* Lint...

Commit:
Update tpch_benchmark.cpp (#657)

Bensk1 committed:
Rename ScanType to PredicateCondition (#659)

* Rename ScanType to PredicateCondition

* Format

Commit:
Print column count for operators (#658)

* Print column count for operators

* format

f4lco committed:
Fix set child (#655)

* Call _child_changed after child has been set

"set_child" should call "_child_changed", which
is, well, the internal routine handling changes
in children nodes.
In addition to the current behavior this will
add the following improvements:
(1) State which was invalidated by the change in
the children node such as statistics
and output column references will now be
cleared and recomputed when required.
(2) "_child_changed" will also call
"_on_child_changed" which applies changes to the
node state specific to the type of node at hand.

* Fix derive_statistics_from on predicate nodes

The column IDs should be retrieved from the passed-in
children and not from the node itself since

(1) the resulting column IDs may not exist or give
wrong results if passed-in-child and this-child
do not agree on a common source of information
(e.g., stored table node).

(2) retrieving the column-IDs may fail since
this-child might not be present in case this
predicate node was detached from the LQP.
This can occur as part of the predicate
reordering optimizer rule.

Bensk1 committed:
Add optimizer rule to create IndexScans (#661)

* Commit to transfer to pella

* Make initial version work

* LQPTranslation

* LQPTranslatorTest

* Between

* Minor

* Format

* ScanType -> PredicateCondition in comments

* Rename scan_typee

* Some cleanup

* Some refactoring

* Add test

* Minor

* Lint + Format

* Address review issues

* Add IndexInfo type

adi64 committed:
Fix omitted lazy column statistics initialization (#669)

* Fix omitted lazy column statistics initialization

* rename TableStatistics::column_statistics to
TableStatistics::get_all_column_statistics, make it non-const (as it
lazily generates ColumnStatistics)

* add test for TableStatistics::get_all_column_statistics

* formatting (oops)

* Change back to TableStatistics::column_statistics

* make TableStatistics::column_statistics const (also make the underlying
member mutable and other lazy getters const)

adi64 committed:
Add method to remove an index from a chunk (#662)

* implement Chunk::remove_index

* format

* Replace conditional Fail() with DebugAssert

* add AddIndex test for Chunk

* add tests for index methods of Chunk

* fix test names. whoops :D

Commit:
Memory Usage Estimations for Tables and underlying structures (#664)

* Memory usage

* Format

* reset tpch benchmark

* tests

* include Mvcc members

* tests

* add snakeoil warning

* lqp visualizer

* integrate review

mjendruk committed:
Encoded Column Framework v2 (#644)

* Implement first working imperative packing algorithm

* Implement recursive optimised and templated version

* implement unpack routine

* Fix bugs in packing code

* Add null suppression encodings (encoders + vectors)

* Fix compiler errors

* Fix SimdBp128Encoder

* Add decoders

* Fix compiler errors

* Extract SIMD-BP128 algorithm into separate class

* Add decode method to NsDecoder

* Fix compiler errors simd_bp128_decoder

* Implement SimdBP128Decoder get method

* Rename BaseAttributeEncoder to BaseNsEncoder; fix decoding bug

* Rename BaseEncodedVector to BaseNsVector

* Add NsType enum and utility functions

* Update AttributeVectorIterable and DictionaryIterable; format code

* A lof of bug fixes and simplifications; fix ReferenceColumnIterable

* Format code

* Fix duplicate constructor

* create duplicates of new dictionary column classes; refactor null suppression interface

* Revert dictionary column classes to master

* Duplicate base_dictionary_column.hpp and revert to master

* Update operators/table_scan from master

* Fix compiler errors

* Add new dictionary column and make it compile

* Implement lots of conceptual things; refactoring

* Fix compiler errors; refactor code; add templates; document

* Rename tuples

* Refactoring (on my way to resolving encoded columns)

* Finalize new encoded columns interface

* Implement table scans for encoded columns

* Switch back to old dictionary encoding;

* Reset playground

* Add one more comment

* Format code

* I ❤️ linter

* update CMakeLists.txt

* Add options to choose encoding type in DictionaryCompression class

* Replace __m128i with uint128_t in containers (GCC error)

* Small changes (potential fix)

* Add workaround for auto non-type template parameters

* Fix narrowing conversion

* Potential fix for GCC compiler error

* fix compiler error

* Format code

* add benchmark with new dictionary encoding

* Add tests for SIMD-BP128 and fix bugs

* Add tests for all null suppression types

* move null_suppression_test.cpp

* Update NewDictionaryEncoder (use correct byte size); fix small bug

* Add BaseNewDictionaryColumn and start updating TableScan

* Format code

* Linter why you do dis to me? 😢

* Add more tests

* Fix buffer read overflow

* Run TableScanTest for both dictionary column implementations

* Add documentation; fix bugs

* Add some comments

* Add table scan optimization for new dictionary column

* Fix benchmarks; add unoptimized dictionary column scan (commented out)

* Format code

* Rename EncodingType::Dictionary to ::DeprecatedDictionary and ::NewDictionary to ::Dictionary

* Rename DictionaryColumn to DeprecatedDictionaryColumn

* Rename iterators

* Rename rest

* Format code

* Shorten comment

* Add test for decode()

* Use NsVector::cend() in tests

* renemae folder null_suppression to zero_suppression

* Rename BaseNsDecoder to BaseZeroSuppressionDecoder

* rename BaseNs(Encoder | Vector) to BaseZeroSuppression(Encoder | Vector)

* Correctly name encoder file

* Rename remaining classes and files

* Format code

* Change references to zs_utils.hpp

* Add data_size() to BaseEncodedColumn

* Move data_size() method to BaseColumn interface

* Addition to last commit

* Fix ValueColumn<T>::data_size

* Add more tests for SIMD-BP128

* Fix insufficient coverage checking

* Fix SIMD-BP128 tests

* Add test for 32bit and remove unsupported flags

* Fix some bugs

* Fix simd bp128 encoding for 32 bit size

* Increase test coverage a bit

* Refactor SIMD-BP128 code a bit

* Use BaseDictionaryColumn instead of BaseEncodedDictionaryColumn in visitor pattern. Use same optimisations for both dictionary implementations in TableScan

* Fix BaseDictionaryColumn::visit()

* Remove data_size methods for now

* Format code

* Rename _encode to _on_encode; remove dereference_column.hpp

* Add an explanation

* Remove dereference_column.hpp

* Remove duplicate include

* Add comments

* Update comment

* Fail instead of return nullptr (encode)

* Add semicolon

* Add class comment

* Remove Op prefix

* Fix column encoder template

* Add BaseEncodedColumn to visitor interface

* Add run-length encoding

* Add tests for encoded columns

* Format code

* Remove write_string_representation and copy_value_to_value_column

* Fix test

* Return BaseEncodedColumn instead of BaseColumn (Pick)

* Use passed allocator in all encoders (Pick)

* Shrink RLE vectors to fit (Pick)

* Split ZsType into interface enum (ZsType) and internal enum for vector types (ZsVectorType); update interface of encode_by_zs_type

* Format code

* Remove old comment and add final

* Add ZsType to column encoder interface; add ChunkEncoder

* Format code

* Fix GCC linker error

* Rename EncodingType::Invalid to EncodingType::Unencoded

* Revisit assert message; add additional explanation

* Use aligned SIMD functions if possible

* Dismantle column_encoders folder

* Dismantle encoded_columns folder

* Dismantle iterables folder

* Move files out of storage/

* move certain files into column_iterables; refactor column_value.hpp -> column_iterator_values.hpp

* Rename iterators in value_column_iterable.hpp

* Fix bug (NonNullPointAccessIterator can in fact be null if INVALID_CHUNK_OFFSET

* Move files into column_iterables; Merge encoded_columns + resolve_encoded_column_type; Replace Column*Info with TemplateType

* Replace hana::map with std::map in column_encoding_utils.cpp

* Move zero suppression implementations into separate folders

* Replace hana::map with std::map

* Merge files; add comments

* Store null values in additional bit vector instead of encoded by a magic number

* Add comments

* Add comments

* Add another comment

* Rename zero_suppression/utils.hpp to encode_by_zs_type.hpp; Add some more comments

* Fix merge errors

* Format code

* Sort CMakeLists.txt

* Address PR comments no. 1

* Address PR comments no. 2

* Refactor dictionary_column_iterable.hpp and attribute_vector_iterable.hpp

* Remove detail::all_data_types

* Address PR comments no. 3

* Address PR comments no. 4

* Address PR comments no. 5

* Format code

* Rename DictionaryCompression to DeprecatedDictionaryCompression; other stuff

* Change parameter order

* Change parameter order

* Address PR comments no. 6

* Fix bug

* Rename all handle methods in ColumnVisitable to handle_column; add comment

* Fix clang warning

* Format code

* Merge three files in /zero_suppression

* Rename everything related to zero suppression

* Format code

* Fix merge

* Rename more variables

* Format code

* Fix merge

* Fix merge

* update sql-parser

* jeeeenkins

* Fix tests after merge

* Add wiki link

* Update after merge

* Format code

Commit:
RelWithDebInfo gets LTO and arch=native (#677)

mrks committed:
Fix OSX build, add print for operators (#680)

f4lco committed:
Alias with Column Renames (#676)

* update sql-parser

* Implemented column renames as part of table alias in hyrise

* Adjust to recent changes in sql-parser (alias struct, join types)

lanice committed:
Add Asserts to SQLTranslator when INSERTing mismatch type values (#685)

* Add Asserts to SQLTranslator to prevent INSERT of mismatch types.

* Fix SQLTranslatorTest::InsertValuesColumnReorder.

* Add comments, rename variables.

* Format.

* Add test SQLTranslatorTest::InsertInvalidDataTypes.

* Rename variable.

lawben committed:
Add query cache to pipeline (#666)

* Add query cache to pipeline

* PR comments

* Wrap transaction context

* Fix lint

* Wrap raw pointers

* PR comments

* Fix nullptr access

* PR comments

mrks committed:
CI: Survive lint fail (#690)

mrks committed:
Improve test isolation in physical_query_plan_test.cpp (#694)

mrks committed:
Allow deep copy of diamond LQPs (#691)

* Allow deep copy of diamond LQPs, fixes #625

* address review

* format

Commit:
Custom Optimizer in SQLPipeline (#693)

* itisdone

* things

* did i change something?

* did i change something?

* did i change something?

* chunkusemvcc->usemvcc

* things

* something

* format

lanice committed:
Nested Expressions in WHERE clause + Optimizier Rule (#672)

* Add possibility to have nested Expressions in WHERE clause.

* SQLTranslator: Remove own calculation of nested Expressions, add ProjectionNode instead.

* Adjust to new column origins PR.

* Remove added column expression when having a nested expression (SQLTranslator).

* Format.

* Fix SQLTranslatorTest.ExpressionTest.

* Add first part of NestedExpressionRule to Optimizer for optimizing LQPs containing expressions like 1233+1 [WIP].

* Finish optimizer rule NestedExpressionRule.

* Format.

* Move get_function_from_expression to utils/.

* Format.

* Use LQPColumnReference instead of ColumnID in SQLTranslator for nested expressions in predicates.

* Make NestedExpressionRule more generic, replacing ColumnReferences pointing to calculatable expressions.

* Adjust NestedExpressionTest.

* Format.

* Make NestedExpressionRule a RuleBatchExecutionPolicy::Once rule.

* Add NestedExpressionRuleTest.

* Make Optimizer rule remove Expression columns after replacement.

* Format.

* Fix NestedExpressionRule removing Expression column from Projection.

* Rename some local variables in NestedExpressionRule.

* Use guard clauses in NestedExpressionRule.

* Add comments to NestedExpressionRule.

* Format

* Add test for NestedExpressions with columns.

* Format.

* Fix typo in comment.

* Change NestedExpressionRule to ConstantCalculationRule, fix some comments.

* Refactor ConstantCalculationRule methods.

* Adjust names and comments.

* Enable calculation of Expression NULL+NULL in ConstantCalculationRule.

* Format.

* Fix ConstantCalculationRuleTest after merge.

rs22 committed:
WIP: Use futures for asynchronous operations

rs22 committed:
Code looks terrible, but session startup works again

rs22 committed:
Introduce then operator
It sets the launch policy and gets the future's value
(thereby throwing exceptions)

rs22 committed:
Further clean up code by making then operator smarter

rs22 committed:
Working towards a functioning SimpleQueryCommand

rs22 committed:
Try to fix SimpleQueryCommand

rs22 committed:
Make SimpleQueryCommand work again

rs22 committed:
Clean up server task dispatching

lawben commented on src/lib/server/client_connection.cpp:
```diff
@@ -0,0 +1,139 @@
+  PostgresWireHandler::write_value(*output_packet, NetworkMessageType::SslNo);
+  
+  return _send_bytes_async(output_packet, true)
+    >> then >> [](unsigned long) { /* ignore the result */ };
```
we don't use `unsigned long` in hyrise. let's change all of these to `uint64_t`

lawben commented on src/lib/server/hyrise_server.cpp:
```diff
@@ -18,7 +19,8 @@ void HyriseServer::accept_next_connection() {
   if (!error) {
-    std::make_shared<HyriseSession>(std::move(_socket), _io_service)->start();
+    auto connection = std::make_shared<ClientConnection>(std::move(_socket));
+    std::make_shared<HyriseSession>(_io_service, connection)->start();
```
comment here that this needs to be shared_ptr because we call shared_from_this in start. otherwise a shared_ptr makes no sense

lawben commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -23,278 +23,270 @@
-
-  _response_buffer.insert(_response_buffer.end(), output_packet.data.begin(), output_packet.data.end());
+  
+  (_perform_session_startup() >> then >> [=] () { return _handle_client_requests(); })
```
comment why we use `.then()` explicitly here

lawben commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -23,278 +23,270 @@
-      _send_ready_for_query();
+boost::future<void> HyriseSession::_handle_client_requests() {
+  auto process_request_header = [=](RequestHeader request) {
+    // Process the request header
```
unnecessary comment ;)

lawben commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -23,278 +23,270 @@
+    // Process the request header
+
+    if (request.message_type == NetworkMessageType::TerminateCommand)
+      // Not really unexpected, but this is a way to escape the infinite recursion
```
i don't understand this comment

lawben commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -23,278 +23,270 @@
+      >> then >> process_request_header
+      >> then >> process_command;
+  
+  return command_result.then(boost::launch::sync, [=] (boost::future<void> result) {
```
can we split this up somehow? hard to tell in which case what happens with nested `>> then >>` and then the unwrap and then again...

lawben commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -23,278 +23,270 @@
+  };
+    
+  return create_sql_pipeline() >> then >> [=] (std::shared_ptr<CreatePipelineResult> result) {
+    return result->is_load_table
```
let's use an explicit `if () {} else {}` here, the ternary if with the >> then is confusing

lawben commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -23,278 +23,270 @@
+auto HyriseSession::_dispatch_server_task(std::shared_ptr<T> task) -> decltype(task->get_future()) {
+  using TaskList = std::vector<std::shared_ptr<AbstractTask>>;
+
+  CurrentScheduler::schedule_tasks(TaskList({task}));
```
i think we can omit  `TaskList({task})` and only write `{task}`

rs22 replied:
I've tried -- we can't :(

lawben commented on src/lib/server/then_operator.hpp:
```diff
@@ -0,0 +1,93 @@
+template<class T, char, class O> struct half_apply { T&& lhs; };
+
+template<class Lhs, class Op>
+half_apply<Lhs, '*', Op> operator>>( Lhs&& lhs, make_operator<Op> ) {
```
the `char` template param and `'*'` can be removed as we don't use them

lawben commented on src/lib/server/use_boost_future.hpp:
```diff
@@ -0,0 +1,77 @@
+/**
+ * See the documentation for boost::asio::use_future_t for a usage example.
+ */
+#if defined(BOOST_ASIO_MSVC)
```
we can remove this if case because we don't support windows :)

rs22 replied:
I was wondering how closely we should align the use_boost_future.hpp to boost's use_future.hpp. Changing less would make applying later 'upstream' changes easier -- for now I've just replaced some `std::future`/`std::promise` with `boost::future`/`boost::promise`.

lawben replied:
okay, then just leave it in

lawben commented on src/lib/server/use_boost_future_impl.hpp:
```diff
@@ -0,0 +1,163 @@
+
+} // namespace detail
+
+#if !defined(GENERATING_DOCUMENTATION)
```
what is this?

lawben commented on src/lib/tasks/server/create_pipeline_task.hpp:
```diff
@@ -1,14 +1,24 @@
+struct CreatePipelineResult {
+  std::shared_ptr<SQLPipeline> sql_pipeline;
+  std::optional<std::pair<std::string, std::string>> load_table;
+  bool is_load_table;
```
this is redundant. if the optional is set, this is true

lawben commented on src/lib/tasks/server/execute_server_prepared_statement.hpp:
```diff
@@ -4,12 +4,14 @@
 namespace opossum {
 
-class ExecuteServerPreparedStatement : public ServerTask {
+class SQLQueryPlan;
```
why do we do this compared to just including the respective headers?

rs22 replied:
Something I've learnt in various 3D Point Cloud projects -- systematic forward declarations can reduce compile time significantly by not having to include unnecessary headers

lawben commented on src/lib/tasks/server/send_query_response_task.cpp:
```diff
@@ -63,13 +67,13 @@ void SendQueryResponseTask::_send_row_description() {
   }
 
-  _session->async_send_packet(output_packet);
+  _connection->_send_bytes_async(output_packet, false);
```
the naming of send_bytes_async is wrong, i.e. _ is for private methods

stephde replied:
this method ist not public either. what about protected methods? are they with or without underscore?

lawben replied:
protected == private in hyrise. There are very few real private methods, most of them are protected

rs22 replied:
The real issue here is, that SendQueryResponseTask is a friend class of ClientConnection -- this is a purposefully ugly way of making this work.

The SendQueryResponseTask is the only place where our layering is broken -- and I'm still not entirely convinced that just for the purpose of string concatenation on the scheduler thread, we should open ourselves up to such architectural issues and possibly threading problems (now both the Task and the HyriseSession could potentially manipulate the ClientConnection in parallel).

Ideally, we'd pass the ClientConnection into the HyriseSession as a unique_ptr, then we'd have the following layers:

Server Tasks
+++
HyriseSession
+++
ClientConnection + PostgresWireHandler
+++
Boost Asio Socket

rs22 replied:
I think we could achieve this in one of two ways:
1. Transform the SendQueryResponseTask into a static class, move the remaining Packet-Handling related stuff to ClientConnection (keeping the Chunk and Column Loops in the Task) and send the data on the main thread
2. Think about some sort of materialized string representation of the table (something that we also discussed with Stefan) that is being generated on the scheduler thread and afterwards sent to the network on the main thread

lawben commented on src/lib/tasks/server/send_query_response_task.cpp:
```diff
@@ -150,35 +156,39 @@ void SendQueryResponseTask::_send_command_complete() {
     _send_row_data();
   }
 
-  _send_command_complete();
```
what happened here? ;)

lawben commented on src/lib/tasks/server/send_query_response_task.hpp:
```diff
@@ -5,21 +5,26 @@
  protected:
   void _on_execute() override;
 
+  // TODO: Move these to ClientConnection
```
why?

lawben commented on src/lib/tasks/server/server_task.hpp:
```diff
@@ -1,16 +1,22 @@
-  explicit ServerTask(std::shared_ptr<HyriseSession> session) : _session(std::move(session)) {}
+  explicit ServerTask(){}
+  
+  using result_type = T;
```
we don't use this, so it can be removed

rs22 replied:
It's being used in HyriseSession::_dispatch_server_task

lawben commented on src/lib/tasks/server/create_pipeline_task.hpp:
```diff
@@ -1,14 +1,24 @@
+  bool is_load_table;
+};
+
+// TODO: I think we can make the CreatePipelineResult a unique_ptr again...
```
then let's do it :)

rs22 committed:
Re-add Extended Query Protocol Handling (untested)

rs22 committed:
Fix server task naming

rs22 committed:
Refactor SendQueryResponseTask, fix EQP

rs22 committed:
Enable recreation of Delete operator, fix transaction lifetime

rs22 committed:
Incorporate some feedback

rs22 committed:
Reset playground

rs22 committed:
Formatting

rs22 commented on src/lib/server/client_connection.hpp:
```diff
@@ -0,0 +1,60 @@
+};
+
+class ClientConnection {
+  friend class SendQueryResponseTask;
```
here's that ugly friend class specifier

lawben replied:
ihhhhhhhhhhhhhhh ;)

rs22 commented on src/lib/server/hyrise_session.cpp:
```diff
@@ -8,293 +8,295 @@
+  auto process_command = [=](std::pair<RequestHeader, InputPacket> packet) {
+    switch (packet.first.message_type) {
+      case NetworkMessageType::SimpleQueryCommand: {
+        const auto sql = PostgresWireHandler::handle_query_packet(packet.second);
```
Ideally we wouldn't use the PostgresWireHandler here, that would help writing tests on a higher level later (by exchanging the ClientConnection for a mock).

rs22 committed:
Abort transactions on any exception thrown by a command

rs22 commented on src/lib/server/hyrise_server.cpp:
```diff
@@ -18,7 +19,9 @@ void HyriseServer::accept_next_connection() {
-    std::make_shared<HyriseSession>(std::move(_socket), _io_service)->start();
+    auto connection = std::make_shared<ClientConnection>(std::move(_socket));
+    // The session will retain a shared_ptr to itself as long as it's alive
+    std::make_shared<HyriseSession>(_io_service, connection)->start();
```
What would be a better name for `HyriseServer` and `HyriseSession`?

lawben replied:
just: Server and ServerSession

rs22 committed:
Remove unused template parameter from operator definition

stephde committed:
fix wire_handler tests

stephde committed:
add gmock test for hyrise session startup

julkw committed:
semi and anti join node output column references fix (#701)

* fix output_column_references for Semi Join Node

* test for output column references of semi and anti joins

* correct indentation

* variable renaming

Commit:
Print Acyclic Directed Graphs (#706)

* introduce helper mthd

* that was unnecessary

* format

* unplanned

* jenkins

mjendruk committed:
Remove dictionary compression class #2 (#705)

* Remove DeprecatedDictionaryCompression and replace it with ChunkEncoder

* remove duplicate include

Commit:
Fix JoinHash/MurmurHash with different types (#653)

* Fix JoinHash with different types

* Disabled mixed type join test for SMJ

* Improved mixed type joins for JoinHash

* Remove unused JoinHashTraits

* Rework traits with PR feedback

* lint

* fixed width integer type

* Use type_cast. Rename HashType.

jfrohnhofen committed:
JIT 1: adding LLVM and jit flag to cmake files  (#696)

* JIT: added LLVM and jit flag to cmake files

* JIT: fix compiler flags for LLVM generation

* JIT: Review feedback

Commit:
Deep LQP comparison (#673)

* Jenkins where are you

* current state

* compiles

* whateva

* current state

* thisisit

* lol

* consider it a prove of concept

* Look how nice the predicate reordering rule test looks

* all is violent, all is green

* everything passes

* lintylint

* current state

* things

* remove make

* tmp

* format

* whateva

* AbstractLQPNodeCRT

* current state

* works

* lintylove

* changed some name

* PR comments

* PR comments

* disable make for CreateViewNode

* compiles

* sum changes

* foo

lawben committed:
Small fixes

lawben committed:
Merge upstream master

lawben committed:
Remove old lines from pgtest

mjendruk committed:
Fix SIMD-BP128 decompressor (#720)

* Add failing test

* Fix tests

* Fix integer overflow in simd bp128 decompressor

* Remove assertion

* Format code

* Lint

lawben committed:
Small changes

Commit:
Use int64_t instead of long (#724)

* Use int64_t instead of long

* Satisfy linter

mjendruk committed:
Replace bimap with unordered_map (#726)

lawben committed:
Fix things and linter

lawben committed:
Remove old unused SQLQueryOperator (#734)

* Add query cache to pipeline

* PR comments

* Wrap transaction context

* Fix lint

* Wrap raw pointers

* PR comments

* Fix nullptr access

* Remove SQLQueryOperator

* PR comments

* Delete unsued files

* Small merge fix

* CMake format fix

Commit:
JoinGraph and JoinGraphBuilder (#708)

* Checkout files from dev branch

* <3 lint

* JoinEdge comment

* JoinGraph comment

* join edge comment

* address some pr comments

* converter became a builder

* mwm

* lint happy

* no more print

* lol

* comment

* jenkins

lawben committed:
Merge branch 'master' into futuristic_eqp

rs22 committed:
Try to work around CI compilation problem

rs22 committed:
Revert "Try to work around CI compilation problem"

This reverts commit 1d752005b73c45d77cdaefe451417c5a5b041fdd.

rs22 committed:
Add use_boost_future implementation for boost 1.63

rs22 committed:
Add boost libraries as linker dependencies for test targets

rs22 committed:
Remove unnecessary semicolons

stephde committed:
clean up benchmark script

jfrohnhofen committed:
JIT 2: types, type dispatching, basic operations (#709)

* JIT: added LLVM and jit flag to cmake files

* JIT: fix compiler flags for LLVM generation

* JIT: Review feedback

* JIT: added jit types, type dispatching and computation framework

* JIT: fix compilation

* JIT: lint love

* JIT: trigger rebuild

* JIT: add bool type to opossum

* JIT: PR2 feedback

* JIT: add comment about JitVariantVector

* JIT: add Bool to DataType, but not to DATA_TYPE_INFO

* JIT: fixed set_is_null

* JIT: clang-format

* JIT: removed unnecessary include

* JIT: PR2 feedback

lawben committed:
Add PreparedStatement support in SQLPipeline (#732)

* WIP prepared statements

* Add tests

* Move stuff in pipeline

* Add prepared statements to console

* Add correct number of parameters

* Add INSERT/UPDATE support

* Add test for DELETE

* PR comments

* Fix cache set error

mjendruk committed:
Apply type erasure on iterators in order to greatly speed up compilation in debug mode (#704)

* Add AnyColumnIterable used to erase type of iterators returned by iterables; other fixes and small improvements

* I ❤️ our linter

* Add virtual destructor to AnyColumnIteratorWrapperBase

* Add documentation and test

* Fix code; make AnyColumnIterator’s constructor private in order to prevent it being used outside of AnyColumnIterable

* Address PR comments

* Format code

jfrohnhofen committed:
JIT: PR feedback 2.1 (#738)

* JIT: PR feedback 2.1

* force rebuild

* JIT: undo sqllike_to_regex change

Commit:
Fix Self Product (#739)

Commit:
Rename child/parent in LQP to input/output (#743)

* started

* continues

* more

* format

Commit:
Remove SQLPlanner (#744)

* Remove SQLPlanner

* I actually managed to screw this up

Commit:
Format duration (#747)

* Format duration

* doc

Commit:
Only evaluate Assert msg on failure (#746)

* Only evaluate Assert msg on failure

* fixyfix

* fixyfix

* remove Assert template

rs22 committed:
Add session startup tests

lawben committed:
Merge master

lawben committed:
First PR comments addressed

Commit:
Column type independent materialize() (#750)

* woaks

* tests

* lintylove

* doc

* address PR

* doc

* doc

* doc

* doc

* doc

j-beyer committed:
Add index join implementation (#667)

* A first implementation of the index join, untested as of yet

* formatted

* JoinIndex now conforms to join interface

* a simple test for JoinIndex

* formatting

* Adding a caveat comment

* Making the index join work with nested loop join as a fallback.
Now supports all join types.

* Cleaning up todos

* Incorporate review into join_index_test.cpp

* Incorporate review into join_index.cpp

* switched around the right/left chunk iteration order for cached indices
removed compatability fallback for unsupported join types

* removed switching of columns

* enabling right and outer joins

* there are no further index types

* fixing left matches

* adding more tests for outer joins

* larger test sets

* fixed left join for multi chunk case

* formatting

* index join works correctly even without an index

* more tests (currently missing groupKeyIndex, it is still having trouble building correctly)

* fixing comparison order to match other join operators

* using explicit comparisons to ensure correctness and expressiveness

* removing unneeded variable

* clarifying comments

* refactoring variable order for consistency

* Clean up comments and TODO

* fixing typo

* testing on all common join tests

* Tests that specifically test with indexes

* Reformat, restructure

* cleaning up unused variable

* updating to build with new master

* disabling test that crashes due to AdaptiveRadixTreeIndex not behaving to spec

* referencing issue for disabled test

* formatting

* basic benchmark harness for index join comparison

* Add table generator method for generating skewed normal distributions and pareto distributions

* Format and lint

* more benchmark work for index_join

* set left_matches only when needed

* const constable variables

* cleaning up new table generator interface and implementation

* more work on benchmarking and table generator

* updating index join to eagerly allocate memory

* Incorporate review of #667

* fixing double initialization of _left_matches

* formatting and linting

* fixing more typing lints

* Fix merge errors and adapt new table generator method to also use EncodingType instead of compress boolean

* Apply table generator review

* More review changes:
* fix left_matches and _right_matches having different depths
* avoid repeated switch in table generation by wrapping value generation in std::function
* use number as column name instead of increasing characters from 'a'
* comments and variable names

* Use ChunkEncoder instead of DictionaryCompression in index join test

* Use DeprecatedDictionaryCompression instead of ChunkEncoder: not all indices support chunk encoder yet

* Format, add todo

* Use new slim encode_all_chunks api

* Split up _join_two_columns into index join and nested loop join implementations

* Use DeprecatedDictionary encoding type in join benchmark; add todo to replace

* Incorporate review feedback

* Pull up index join further

mjendruk committed:
Update binary importer and exporter to support new dictionary column instead of old (#754)

rs22 committed:
Some more tests, comments and clean up

rs22 committed:
SimpleQueryCommand test, introduce TaskRunner

rs22 committed:
Merge remote-tracking branch 'origin/futuristic_eqp' into futuristic_eqp_tests

rs22 committed:
EQP test, move packet content parsing to ClientConnection

rs22 committed:
Add more tests, improve error messages

rs22 committed:
Fix test

Bouncner committed:
Add missing sstream include (#757)

rs22 committed:
WIP: Integrate SendQueryResponseTask into Session thread

Commit:
Add TPCH 22 query (#762)

* Add TPCH 22 query

* Remove legacy variable

mjendruk committed:
Support encoding types in csv exporter (#753)

* Support all encoding types in ExportCsv

* Add test

Commit:
Fix query in SQLBenchmark (#763)

* Fix query in SQLBenchmark

* lalala

rs22 committed:
Add missing pieces for sending query responses

tbjoern committed:
Chunk Pruning for Predicate Nodes based on Pruning Filters (#668)

* add *.txt.user to gitignore

* add ChunkStatistics that get recorded during Chunk compression

* run clang-format

* add chunk_statistics to chunk

* implement primitive/first version of chunk pruning rule

* ChunkStatistics(can_prune): handle AllTypeVariants properly through template subclasses, adapt tests to conform to fixed behavior

* ChunkStatistics: implement more operators

* reorder parameters + add DebugAssert

* simplify simple test

* add intersection test to pruning (still fails)

* fix(chunk_pruning_rule): fix predicate chain detection

* test(chunk_pruning): add test for edge cases of comparators

* merge(master): resolve merge issues

* get_table: implement exclusion list handling

add test for it in chunk_pruning_test

* add script for full tpch headers

* conduct first benchmarking test

chunk_pruning_rule: skip over validation nodes
add to_string to chunk_statistics
change console load table to compress chunks and have chunk size 500000

* add GetTable test for excluded chunks

* GetTable: move pruning description from name() to description()

* fix rebase issues, fix segfault when there were no chunk statistics available for a chunk, add test for the latter

* apply linter and format, fix get table description

* run format

* define default chunk size for console load table

* add comment to nested ternary for child count

* add BOOST_NO_AUTO_PTR because clang removed auto_ptr for c++17

* use optional instead of experimental/optional

* address review comments (remove code, add comments, refactor)

* remove insert header script

* add comments in GetTable on_execute to explain chunk pruning

* move pruning rule to the final batch in the optimizer to only run it once

* add clarifying comment for always returning false in pruning

* refactor chunk statistics - seperate filters into seperate classes

* start implementing range filter generation

* first fixes of range filter

* implement and/or fix RangeFilter + add tests + add new table

* add chunk stats calculation onto new encoding system

* adjust comment

* fix compiler errors, copy over statistics generation for deprecated dictionary columns

* address review comments

* fix merge errors

implement some suggestions
check for variant in exclusion list computation via if-statement, instead of using DebugAssert

* get_table: use unordered set with excluded chunks for calculation of pruned table

* abstract_lqp_node: use std::count in input_count instead of ternary operator

* chunk_pruning_rule: remove boost/variant include
chunk_statistics::MinMaxFilter: make member variables const

* chunk_statistics: use const for t_value,
declare destructors with override instead of virtual,
add inline documentation for BaseFilter::can_prune

* chunk_statistics: move max_ranges_count into constexpr

* lqp node: remove is_leaf function

* get_table: rename stringstream variable

* chunk_statistics: move build statistics methods to cpp

* run format

* run linter

* get_table: make excluded chunks a simple vector
console: mark constant constexpr

* chunk statistics cpp: manual format

* try to fix structured binding compiler error

* rename base filter to abstract filter

* chunk statistics: split filter and column statistics into seperate files

* run format

* move chunk_statistics code to seperate folder

* add documentation to filter classes

* add doc to chunk statistics

* run format and lint

* implement suggestions

* format + lint

* implement more suggestions, fix failing tests

expand constructors of get table

* remove unnecessary optional includes, fix typo, clarify abstract_filter description, re-add deleted comment in abstract lqp node

* Addressed Moritz' review, harmonized interface with TableScan

* Revert "Merge branch 'feature/chunk_statistics' of https://github.com/apopiak/hyrise into apopiak-feature/chunk_statistics"

This reverts commit 29aacf60c19b844ad3698cd2cb9c15ce955888a7, reversing
changes made to 2eb62213b03e7a1457f4b6692e83b0a72dc89932.

* Revert "Revert "Merge branch 'feature/chunk_statistics' of https://github.com/apopiak/hyrise into apopiak-feature/chunk_statistics""

This reverts commit ff2cac48661c67dc38dbb8c1634908a05d56afd2.

* chunk_column_statistics: add general method to capture new encoded column types

rs22 committed:
Clean up

lawben committed:
PR comments

rs22 committed:
Fix segfault when sending query results

lawben committed:
Fix type

lawben committed:
Fix all the things

rs22 committed:
Add some comments

lawben committed:
Fix more things

Commit:
Fix PQP/LQP Visualizers for diamond shapes (#769)

* Did things

* format

Commit:
Fix AbstractOperator::recreate() for diamonds and diamond Task plans (#761)

* right now

* tests

* format

* varname

* Spelling

mjendruk committed:
Use new dictionary column in indices (#765)

mrks committed:
CI: Only build sanitizer for master (#772)

* bla

* bla

* gla

* bla

* gae

* bla

* logall

* gaesga

* bla

* bla

* ccache log raus

* agwetg

* awgeg

* tatat

mrks committed:
Make visualization background black (#773)

Commit:
Optimizer predicate pushdown rule (#729)

* Bootstrap predicate pushdown rule

* Implement predicate pushdown for inner joins

* added 3 basic join optimization tests

* add SortNode to whitelist + test

* Add complex test case for pushdown with complex predicates

* format & lint

* some renaming and clean up

* Missing child->input renames

mjendruk committed:
Remove deprecated dictionary column (#766)

Commit:
Faster JoinHash output assembly (#760)

* actually passes tests

* docs

* format

* ref counting back in

* addressed

* doc

Commit:
Rigid Table column definitions  (#736)

* Quick experiment

* lulz

* current state

* cs

* progress

* current state

* current state

* something

* current state

* current state

* things work right now

* Works even nicer

* all build

* format

* linty?

* nananana

* some moving

* things

* running

* format

* correct

* fix something

* now?

* all

* now

* Reliance on DebugAssert has to stop

* fix tpcctablegen

* lint

* fix empty input for aggregate

* fix

* now

* PR

* Max gets his pmr_vector

* format

* fix merge

* format stuff

* spelling

* format

* fix csv parser

* jenkins

Commit:
Add support for subqueries in SELECT and WHERE clause (#727)

* WIP: second version of implementing subselects for WHERE

* Add todos

* start on executing subqueries in projection operator

* Working version of subqueries in WHERE clauses (excluding IN)

* Reference column of subselect if input table type is reference

* Fix subselect expressions in projection.cpp after merge

* Allow subselects in SELECT list

* Improve naming of subselect columns

* add IN condition with subqueries

* Remove deprecated include

* Code review

* fix output_column_references for Semi Join Node

* fix duplication problem when using IN

* test for output column references of semi and anti joins

* fix IS NULL segfault

* Move subselect node from AbstractExpression to LQPExpression

* Add support for subqueries in to_string()

* add subselect test for the sql_translator

* write test for and refactor IN command

* Assert that subselects besides IN only return onw row

* Add test for translation of projection node with subselect

* Add test that assures projection correctly executes subqueries

* Add test for subqueries that return multiple rows

* Give different error, if there are no subselect results

* add queries to sqlite testrunner to test subqueries

* Fix transaction contect being set if it is not available

* Fix inefficient access of the subquery result

* Fix cases where subquery returns a dictionary or reference column

* Add comments

* add subquery test

* Remove unused include

* Run format

* Reset playground

* Remove confusing todo

* Reset playground

* Beautify diff

* Rename child to input to fix merge

* Update sql-parser submodule

* Rename ExpressionType::Select to Subselect and clarify PQPExpression members

* Add comments for lqp expressions

* Format

* Use Table::get_value with disabled performance warnings since we only extract one value

* remove optional around shared pointers

* Fix issue that leads to executing operators multiple times

* Respect aliases for subselects

* Format

* Fix out of bounds access in TableStatistics

* Adjust comments

* Update sqlite_testrunner_queries.sql

* fix merge

apopiak committed:
Implement support for generic BaseEncodedColumn in ChunkColumnStatistics (#767)

* implement dictionary materialization to add generic BaseEncodedColumn support

* add MinMaxFilterTest

* add test that uses a run-length encoded table

* add static assert for range filter

* add range filter test

* use sorted set for dictionary materialization and adjust test so it would fail otherwise

* implement RangeFilter as range of values instead of gaps again

* add string pruning test and only instantiate minmax filter if useful

* address review comments

* clang-format

* fix linting errors

* remove deprecated dict column code, simplify build_statistics code for dict columns

* range_filter: add comment to distances vector

* run format, satisfy linter

rs22 committed:
Merge remote-tracking branch 'hyrise/master' into futuristic_eqp

rs22 committed:
Revert accidentally committed changes to coverage.sh

rs22 committed:
Fix merge

Commit:
BaseColumn::data_type() (#781)

* started

* woaks

* tests

* lintylove

* doc

* address PR

* doc

* doc

* doc

* doc

* doc

* lol

* format

* Basic Case tests pass

* Remove all things case

* remove from CMakeLists

* materilize.cpp gone

* materilize.cpp gone

* materilize.cpp gone

* chunk.hpp

* dedup

mjendruk committed:
Add Frame-of-Reference Encoding (#774)

Commit:
TableStatistics: TableType and column DataType  (#788)

* TableStatistics: TableType and column DataType

* format

lawben committed:
Recreate query plan for subselects (#779)

* Recreate query plan for subselects

* less lines

* format

* Move tests

Commit:
AbstractOperator::type() to query concrete type  (#790)

* AbstractOperator::type() to query concrete type

* lol

* Update table_scan.cpp

rs22 committed:
Remove using directives for boost::asio::ip::tcp

rs22 committed:
Merge remote-tracking branch 'hyrise/master' into futuristic_eqp
